Add `origin key list` subcommand

asymmetric opened the pull request:
This subcommand provides an easy way for users to list the origin keys installed on their system.

Closes #2559.

raskchanky committed:
Add additional logging to LogForwarder and LogPipe.

Right now, there are intermittent problems where the worker will appear to hang
during a large build. The symptoms for this are a build log that freezes and
never progresses past a certain point. Current theories point to zeromq but
replicating the problem has proven difficult.

This change adds additional logging around what's being sent when and also
captures zeromq errors into the logs. Note that these logs are intentionally
left as files and are not sent to jobsrv. If it turns out that zeromq is indeed
part of the issue then our logging of the problem necessarily needs to not be
sent over zeromq.

Signed-off-by: Josh Black <raskchanky@gmail.com>

raskchanky committed:
Add "hab pkg channels" to list channels for a given package.

Signed-off-by: Josh Black <raskchanky@gmail.com>

Commit:
Add `origin key list` subcommand to cli

This subcommand provides an easy way for users to list the origin keys
installed on their system.

Closes #2559.

Signed-off-by: Lorenzo Manacorda <lorenzo@kinvolk.io>

tashimi committed:
Add promote & demote to docs 

Signed-off-by: Tasha Drew <tasha.drew@gmail.com>

chrskhl reviewed (changes requested):
Please update the patch to have a proper commit messages. For example, instead of "first stab at a functioning `list` subcommand" How about "Add list subcommand to cli".

Also, a more detailed description is almost always welcome.

krnowak commented on components/hab/src/command/origin/key/list.rs:
```diff
@@ -0,0 +1,29 @@
+        let path = entry.path();
+
+        if path.is_file() {
+            println!("{:?}", path.file_name());
```
I think you wanted to use `{}` instead `{:?}`, because the latter prints the debugging details, no? Anyway, I wonder if you shouldn't use `common::ui::UI` for printing stuff on the terminal.

chrskhl replied:
How does the rest of the code do this?

krnowak replied:
I still need to find a command that lists something.

krnowak replied:
Well, they seem to use `println!` too. So ignore this please.

asymmetric replied:
I'm actually gonna look into using `UI`, since they use it in the other sibling subcommands.

asymmetric replied:
RE: `{:?}`, if I use `{}` I get `the trait

```
std::fmt::Display` is not implemented for `std::ffi::OsString`
```

krnowak commented on components/hab/src/command/origin/key/list.rs:
```diff
@@ -0,0 +1,29 @@
+pub fn start(cache: &Path) -> Result<()> {
+    for entry in fs::read_dir(cache)? {
+        let entry = entry?;
+        let path = entry.path();
```
You probably could shorten it to `let path = entry?.path()`.

krnowak commented on components/hab/src/main.rs:
```diff
@@ -237,6 +238,10 @@ fn sub_origin_key_import(ui: &mut UI) -> Result<()> {
 }
 
+fn sub_origin_key_list() -> Result<()> {
+    command::origin::key::list::start(&default_cache_key_path(Some(&*FS_ROOT)))
```
I won't pretend I understand this `&*FS_ROOT` thing. ;)

chrskhl reviewed (commented):
Btw, this also looks like a good candidate for squashing the commits. If your just making an improvement to the previous commit and it has not yet been pushed you should do the job of merging those.

asymmetric commented:
@blixtra In this case, I left the two commits to allow @krnowak to review, and I was planning on squashing for the public PR. WDYT?

blixtra commented:
> @blixtra In this case, I left the two commits to allow @krnowak to review, and I was planning on squashing for the public PR. WDYT?

👍

asymmetric commented on components/hab/src/command/origin/key/list.rs:
```diff
@@ -22,7 +22,9 @@ pub fn start(cache: &Path) -> Result<()> {
 
         if path.is_file() {
-            println!("{:?}", path.file_name());
+            if let Some(name) = path.file_name() {
```
Basically `file_name()` [returns an `Option`](https://doc.rust-lang.org/stable/std/path/struct.Path.html#method.file_name).

asymmetric commented:
@krnowak do you think it would make more sense to use [`glob`](https://doc.rust-lang.org/glob/glob/fn.glob.html)?

It would remove the need for the `if let` and would allow us to easily filter only `*.pub`.

krnowak commented on components/hab/src/command/origin/key/list.rs:
```diff
@@ -22,7 +22,9 @@ pub fn start(cache: &Path) -> Result<()> {
         if path.is_file() {
-            println!("{:?}", path.file_name());
+            if let Some(name) = path.file_name() {
+                println!("{:?}", name);
```
Maybe you could use this: https://doc.rust-lang.org/stable/std/path/struct.PathBuf.html#method.display

krnowak commented on components/hab/src/command/origin/key/list.rs:
```diff
@@ -0,0 +1,31 @@
+        let path = entry.path();
+
+        if path.is_file() {
+            if let Some(name) = path.file_name() {
```
Maybe you could use this: https://doc.rust-lang.org/stable/std/path/struct.PathBuf.html#method.display

krnowak commented:
@asymmetric: Makes sense if the directory can have also other files that are not keys. You probably would still be dealing with `PathBuf`s, so the hint about the `display()` method can be still useful.

mwrock committed:
fix mac and windows builds

Signed-off-by: Matt Wrock <matt@mattwrock.com>

reset committed:
Merge pull request #2699 from habitat-sh/fix_mac

fix the mac and windows build

mwrock committed:
bump dev version

Signed-off-by: Matt Wrock <matt@mattwrock.com>

thesentinels committed:
Merge pull request #2705 from habitat-sh/dev_bump

Approved by: @Nobody from Nowhere
Merged by: The Sentinels

reset committed:
Merge pull request #2702 from habitat-sh/promote-demote-docs

Add promote & demote to docs

reset committed:
Default to unstable channel on package upload

Packages should be uploaded to the unstable channel by default and not
the stable channel. The stable channel can be specified as the destination
*as well as unstable* by passing `--channel stable` to the command

Signed-off-by: Jamie Winsor <jamie@vialstudios.com>

thesentinels committed:
Merge pull request #2706 from habitat-sh/upload-unstable

Approved by: @Nobody from Nowhere
Merged by: The Sentinels

mwrock committed:
blog about habitat on windows

Signed-off-by: Matt Wrock <matt@mattwrock.com>

mwrock committed:
fixes #2701 making hab pkg build -w work on windows

Signed-off-by: Matt Wrock <matt@mattwrock.com>

mwrock committed:
deploy scripts should upload to stable

Signed-off-by: Matt Wrock <matt@mattwrock.com>

thesentinels committed:
Merge pull request #2709 from habitat-sh/pkg_build

Approved by: @Nobody from Nowhere
Merged by: The Sentinels

reset committed:
Merge pull request #2710 from habitat-sh/upload_stable

deploy scripts should upload to stable

Commit:
Appease Middleman, fix a few textual & formatting things

Signed-off-by: Christian Nunciato <cnunciato@chef.io>

mwrock committed:
release 0.25.1

Signed-off-by: Matt Wrock <matt@mattwrock.com>

mwrock committed:
fixing a format break

Signed-off-by: Matt Wrock <matt@mattwrock.com>

thesentinels committed:
Merge pull request #2711 from habitat-sh/0.25.1-patch

Approved by: @Nobody from Nowhere
Merged by: The Sentinels

thesentinels committed:
Merge pull request #2708 from habitat-sh/win-blog

Approved by: @Nobody from Nowhere
Merged by: The Sentinels

thesentinels committed:
Merge pull request #2712 from habitat-sh/format

Approved by: @Nobody from Nowhere
Merged by: The Sentinels

mwrock committed:
bumping dev version

Signed-off-by: Matt Wrock <matt@mattwrock.com>

thesentinels committed:
Merge pull request #2713 from habitat-sh/dev_bump

Approved by: @Nobody from Nowhere
Merged by: The Sentinels

asymmetric commented on components/hab/src/command/origin/key/list.rs:
```diff
@@ -0,0 +1,28 @@
+use error::Result;
+
+pub fn start(cache: &Path) -> Result<()> {
+    if let Some(pattern) = cache.join("*.pub").to_str() {
```
What can happen here is that `to_str` returns `None` if the string wasn't valid UTF8.

What I'm doing is I'm ignoring that case, i.e. if it happens, we don't produce an error. Instead, we just print nothing.

WDYT @krnowak ?

krnowak replied:
Sounds fine. I don't think that these filenames have characters outside ASCII set.

asymmetric commented:
@krnowak PTALA, using `glob` now so that we only match `*.pub` files.

thesentinels committed:
Merge pull request #2644 from habitat-sh/jb/more-worker-logging

Approved by: @Nobody from Nowhere
Merged by: The Sentinels

thesentinels committed:
Merge pull request #2704 from habitat-sh/jb/hab-pkg-channels

Approved by: @Nobody from Nowhere
Merged by: The Sentinels

krnowak commented:
Looks good, I think you can file the PR to upstream. If you do, then remember please to close this one.
